src/platform/Zephyr/DiagnosticDataProviderImpl.cpp

```diff
@@ -52,16 +52,48 @@ const size_t kMaxHeapSize = CONFIG_SRAM_BASE_ADDRESS + KB(CONFIG_SRAM_SIZE) - PO
 
 #endif
 
+#ifdef CONFIG_SOC_SERIES_NRF54HX
+#include <hal/nrf_resetinfo.h>
+#endif
+
 namespace chip {
 namespace DeviceLayer {
 
 namespace {
 
 BootReasonType DetermineBootReason()
 {
-#ifdef CONFIG_HWINFO
+#if defined(CONFIG_SOC_SERIES_NRF54HX) || defined(CONFIG_HWINFO)
     uint32_t reason;
+#endif
+
+#ifdef CONFIG_SOC_SERIES_NRF54HX
+    reason = nrf_resetinfo_resetreas_global_get(NRF_RESETINFO);
+
+    if (reason & RESETINFO_RESETREAS_GLOBAL_RESETPORONLY_Msk)
+    {
+        return BootReasonType::kBrownOutReset;
+    }
+
+    if (reason & RESETINFO_RESETREAS_GLOBAL_DOG_Msk)
+    {
+        return BootReasonType::kHardwareWatchdogReset;
+    }
+
+    if ((reason & (RESETINFO_RESETREAS_GLOBAL_RESETPIN_Msk | RESETINFO_RESETREAS_GLOBAL_RESETPOR_Msk)) ==
+        (RESETINFO_RESETREAS_GLOBAL_RESETPIN_Msk | RESETINFO_RESETREAS_GLOBAL_RESETPOR_Msk))
+    {
+        return BootReasonType::kPowerOnReboot;
+    }
 
+    if ((reason & (RESETINFO_RESETREAS_GLOBAL_RESETPOR_Msk | RESETINFO_RESETREAS_GLOBAL_SECSREQ_Msk)) ==
+        (RESETINFO_RESETREAS_GLOBAL_RESETPOR_Msk | RESETINFO_RESETREAS_GLOBAL_SECSREQ_Msk))
+    {
+        return BootReasonType::kSoftwareUpdateCompleted;
```

Comment:
Hmm, how do we know it was caused by the SoftwareUpdate, and not for example crash, or reset from the application layer?

Reply:
I performed several manual tests, and checked what are the values returned from RESETINFO in different cases, for example:

- reset after DFU over Matter -> returned value: 528
- reset using button on board -> returned value:514
- reset from the application -> returned value: 0
- reset from disabling power supply -> returned value: 1
                                                                                                             
Events has unique values, and thanks to it, it can be parsed to actual boot reason
But, I don't know what is returning when board is crashing

Reply:
Right, but I `RESETINFO_RESETREAS_GLOBAL_RESETPOR_Msk | RESETINFO_RESETREAS_GLOBAL_SECSREQ_Msk` does not seem to mean that Firmware Upgrade has been done - and the event to Matter core might be generate incorrectly.

We need to either find out if `suit` module might indicate it, or find some register/retention RAM to store information about reset being scheduled due to image update, not for other reason.

@SylwesterKonczyk @tomchy maybe might help if we have anything like this in SUIT, else we need to look at different way.